coderd/workspaceusage/tracker.go

```diff
@@ -0,0 +1,229 @@
+	BatchUpdateWorkspaceLastUsedAt(context.Context, database.BatchUpdateWorkspaceLastUsedAtParams) error
+}
+
+// Tracker tracks and de-bounces updates to workspace usage activity.
+// It keeps an internal map of workspace IDs that have been used and
+// periodically flushes this to its configured Store.
+type Tracker struct {
+	log       slog.Logger      // you know, for logs
+	flushLock sync.Mutex       // protects m
+	m         *uuidSet         // stores workspace ids
+	s         Store            // for flushing data
+	tickCh    <-chan time.Time // controls flush interval
+	stopTick  func()           // stops flushing
+	stopCh    chan struct{}    // signals us to stop
+	stopOnce  sync.Once        // because you only stop once
+	doneCh    chan struct{}    // signifies that we have stopped
+	flushCh   chan int         // used for testing.
+}
+
+// New returns a new Tracker. It is the caller's responsibility
+// to call Close().
+func New(s Store, opts ...Option) *Tracker {
+	hb := &Tracker{
+		log:      slog.Make(sloghuman.Sink(os.Stderr)),
+		m:        &uuidSet{},
+		s:        s,
+		tickCh:   nil,
+		stopTick: nil,
+		stopCh:   make(chan struct{}),
+		doneCh:   make(chan struct{}),
+		flushCh:  nil,
+	}
+	for _, opt := range opts {
+		opt(hb)
+	}
+	if hb.tickCh == nil && hb.stopTick == nil {
+		ticker := time.NewTicker(DefaultFlushInterval)
+		hb.tickCh = ticker.C
+		hb.stopTick = ticker.Stop
+	}
+	return hb
+}
+
+type Option func(*Tracker)
+
+// WithLogger sets the logger to be used by Tracker.
+func WithLogger(log slog.Logger) Option {
+	return func(h *Tracker) {
+		h.log = log
+	}
+}
+
+// WithFlushInterval allows configuring the flush interval of Tracker.
+func WithFlushInterval(d time.Duration) Option {
+	return func(h *Tracker) {
+		ticker := time.NewTicker(d)
+		h.tickCh = ticker.C
+		h.stopTick = ticker.Stop
+	}
+}
+
+// WithFlushChannel allows passing a channel that receives
+// the number of marked workspaces every time Tracker flushes.
+// For testing only and will panic if used outside of tests.
+func WithFlushChannel(c chan int) Option {
+	if flag.Lookup("test.v") == nil {
+		panic("developer error: WithFlushChannel is not to be used outside of tests.")
+	}
+	return func(h *Tracker) {
+		h.flushCh = c
+	}
+}
+
+// WithTickChannel allows passing a channel to replace a ticker.
+// For testing only and will panic if used outside of tests.
+func WithTickChannel(c chan time.Time) Option {
+	if flag.Lookup("test.v") == nil {
+		panic("developer error: WithTickChannel is not to be used outside of tests.")
+	}
+	return func(h *Tracker) {
+		h.tickCh = c
+		h.stopTick = func() {}
+	}
+}
+
+// Add marks the workspace with the given ID as having been used recently.
+// Tracker will periodically flush this to its configured Store.
+func (wut *Tracker) Add(workspaceID uuid.UUID) {
+	wut.m.Add(workspaceID)
+}
+
+// flush updates last_used_at of all current workspace IDs.
+// If this is held while a previous flush is in progress, it will
```

Comment:
re multi-instance model: this won't overwrite any activity posted on another instance, right?

Reply:
It might, but isn't that the point? Any updates to `last_used_at` posted by another instance will be neccessarily older as we use the flush time as `last_used_at`.

Reply:
I'm thinking about the case when old updates queued on another instance were not sent to DB due to a transient network issue, and after some time the network problem is gone, so the instance will flush and overwrite activity. I guess it is a matter of adjusting the DB query.

Reply:
Something like adding a `WHERE last_used_at < @new_last_used_at`?

Reply:
Yes, unless it is already there.

Reply:
Query updated

Reply:
@mtojek @dannykopping also add an 'integration-style' test with multiple coderd instances.

Reply:
Awesome, LGTM 👍